src/app/clusters/device-energy-management-server/DeviceEnergyManagementCluster.cpp

```diff
@@ -14,438 +16,386 @@
         }
 
     case OptOutStateEnum::kGridOptOut: /* User has opted out from Grid only */
         ChipLogProgress(Zcl, "DEM: OptOutState = kGridOptOut");
         switch (adjustmentCause)
         {
         case AdjustmentCauseEnum::kLocalOptimization:
-            return Status::Success;
+            return DataModel::ActionReturnStatus(Status::Success);
         case AdjustmentCauseEnum::kGridOptimization:
         default:
-            return Status::ConstraintError;
+            return DataModel::ActionReturnStatus(Status::ConstraintError);
         }
 
     case OptOutStateEnum::kOptOut: /* User has opted out from both local and grid */
         ChipLogProgress(Zcl, "DEM: OptOutState = kOptOut");
-        return Status::ConstraintError;
+        return DataModel::ActionReturnStatus(Status::ConstraintError);
 
     default:
         ChipLogError(Zcl, "DEM: invalid optOutState %d", static_cast<int>(optOutState));
-        return Status::InvalidValue;
+        return DataModel::ActionReturnStatus(Status::InvalidValue);
     }
 }
 
-void Instance::HandlePowerAdjustRequest(HandlerContext & ctx, const Commands::PowerAdjustRequest::DecodableType & commandData)
+DataModel::ActionReturnStatus DeviceEnergyManagementCluster::HandlePowerAdjustRequest(const DataModel::InvokeRequest & request,
+                                                                                      TLV::TLVReader & input_arguments,
+                                                                                      CommandHandler * handler)
 {
-    bool validArgs = false;
+    using namespace Commands;
 
-    int64_t power                       = commandData.power;
-    uint32_t durationSec                = commandData.duration;
-    AdjustmentCauseEnum adjustmentCause = commandData.cause;
+    PowerAdjustRequest::DecodableType commandData;
+    ReturnErrorOnFailure(DataModel::Decode(input_arguments, commandData));
 
-    //  Notify the appliance if the appliance hardware cannot be adjusted, then return Failure
-    if (!HasFeature(DeviceEnergyManagement::Feature::kPowerAdjustment))
-    {
-        ChipLogError(Zcl, "PowerAdjust not supported");
-        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
-        return;
-    }
+    ReturnErrorOnFailure(AddResponseOnError(request, handler, CheckOptOutAllowsRequest(commandData.cause)).GetUnderlyingError());
 
-    Status status = CheckOptOutAllowsRequest(adjustmentCause);
-    if (status != Status::Success)
+    DataModel::Nullable<Structs::PowerAdjustCapabilityStruct::Type> powerAdjustmentCapability =
+        mDelegate->GetPowerAdjustmentCapability();
+    if (powerAdjustmentCapability.IsNull())
     {
-        ChipLogError(Zcl, "DEM: PowerAdjustRequest command rejected");
-        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
-        return;
+        ChipLogError(Zcl, "DEM: PowerAdjustmentCapability is Null");
+        return AddResponseOnError(request, handler, Status::ConstraintError);
```

Comment:
```suggestion
        return Status::ConstraintError;
```

and similar everywhere else: returning this in the Invoke will result in an AddStatus